aiida_siesta/workflows/iterate.py

```diff
@@ -0,0 +1,805 @@
+
+from aiida.plugins import DataFactory
+from aiida.engine import WorkChain, while_, ToContext
+from aiida.orm import Str, List, KpointsData, Int, Node
+from aiida.orm.nodes.data.base import to_aiida_type
+from aiida.orm.utils import load_node
+from aiida.common import AttributeDict
+
+from ..calculations.tkdict import FDFDict
+from .base import SiestaBaseWorkChain
+
+class BaseIteratorWorkChain(WorkChain, ABC):
+    '''
+    General workflow that runs simulations iteratively.
+
+    The workchain itself can not be used. To use it, you need to define classes
+    that inherit from it. These classes will have one main job:
+
+    Defining a `run_process` method that runs the process given the current
+    value of the iterator. Note that this method does not need to take care of anything
+    else! Following, we display the expanded outline of the workchain in pseudocode
+    to help you understand how it works and realise what do you want to overwrite:
+
+    -------------------------------------------------------------------------------
+    cls.initialize:
+        cls._get_iterator:
+            keys = [self._parse_key(key) for key in keys]
+
+    while (cls.next_step): # cls.should_proceed and cls.store_next_val are called inside cls.next_step!
+        - cls.run_batch:
+            for i in range(batch_size):
+                if i!= 0:
+                    cls.store_next_val
+                cls.run_process # You have the current value under self.current_val and
+                                # the keys under self.iteration_keys
+                
+        - cls.analyze_batch:
+            for process in batch:
+                cls.analyze_process(process)
+    cls.return_results
+    --------------------------------------------------------------------------------
+
+    You have probably noted that you can analyze a process without worrying about the batch
+    size by overwriting the `analyze_process` method.
+
+    The main idea is that you probably don't need to subclass this class, because there is
+    probably a subclass defined to serve a general purpose that your workflow fits into.
+    See `InputIterator`, for example. The `BaseIteratorWorkchain` goal is just to provide
+    a basis to handle the iterator and the batch size in a consistent way and facilitate
+    further development.
+
+    Finally, the class might seem unnecessarily complex due to the big number of methods
+    defined in it. In reality, it is quite simple. The splitting of each step in separate
+    methods is intentional to help the developer/user overwrite smalls bits of it without
+    having to reimplement the whole step. Therefore: yes, you are invited to overwrite each
+    little functionality to make it useful for your case :)
+
+    You can even overwrite the input serializers. See `_iterate_input_serializer` and
+    `_values_list_serializer`
+    '''
+
+    @classmethod
+    def _iterate_input_serializer(cls, iterate_over):
```

Comment:
> Then a comment. We should be very clear in the documentation that if you want to iterate over some input ports, the user must pass nodes. For instance something like this:
> 
> iterate_over = {"code" : [Siesta@cp1, Siesta@cp2]}
> is not admissible.

Well, this is not admissible because `SiestaBaseWorkChain` does not admit it :) So I think the warning (if needed) should be in `SiestaBaseWorkChain` in this case.

> Finally, we should mention in the documentation that the workchain creates a node for each element in the list, even in the case of normal parameters (no input ports). The user might, in fact, be surprised by the amount on nodes created.

Yes, this is a problem. I believe a better fix might be to think of a smart way to avoid this, then we don't need to warn anyone :) I think it might be quite easy to do if we decide to do it. On the other hand, this values go to parsing functions that maybe in the aiida way of doing things should be `calcfunctions`. If that is the case, then the nodes are going to get created anyway. In your opinion, should parsing functions be `calcfunctions`?

Reply:
> Well, this is not admissible because `SiestaBaseWorkChain` does not admit it :) So I think the warning (if needed) should be in `SiestaBaseWorkChain` in this case.

An AiiDA user knows that it has to pass a node; in your WorkChain it's just a bit confusing because for any other parameter (no inputs ports), you can pass values. But anyway, it's ok! We just have to be clear in the documentation. 

> Yes, this is a problem. I believe a better fix might be to think of a smart way to avoid this, then we don't need to warn anyone :) I think it might be quite easy to do if we decide to do it. On the other hand, these values go to parsing functions that maybe in the aiida way of doing things should be `calcfunctions`. If that is the case, then the nodes are going to get created anyway. In your opinion, should parsing functions be `calcfunctions`?

You are right, if the value is used later on in a `calcfunction`, it is not a problem. The `set_up_parameters_dict` (and other `parsing_functions`) should be a `calcfunction`.? I believe yes. Apparently AiiDA don't complain about it?
Anyway, I will investigate it better, but for the moment is ok like it is.